contracts/services/PayoutManager.sol

```diff
@@ -0,0 +1,214 @@
+        merkleTreePathValidator = _merkleTreePathValidator;
+    }
+
+    //------------------------
+    //  MODIFIERS
+    //------------------------
+    modifier payoutExists(uint256 _payoutId) {
+        require(_payoutId < currentPayoutId, "PayoutManager: payout with specified ID doesn't exist");
+        _;
+    }
+
+    modifier payoutNotCanceled(uint256 _payoutId) {
+        require(payoutsById[_payoutId].isCanceled == false, "PayoutManager: payout with specified ID is canceled");
+        _;
+    }
+
+    modifier payoutOwnerOnly(uint256 _payoutId) {
+        require(payoutsById[_payoutId].payoutOwner == msg.sender, "PayoutManager: requesting address is not payout owner");
+        _;
+    }
+
+    modifier payoutNotClaimed(uint256 _payoutId, address _wallet) {
+        require(payoutClaims[_payoutId][_wallet] == false, "PayoutManager: payout with specified ID is already claimed for specified wallet");
+        _;
+    }
+
+    //------------------------
+    //  READ-ONLY FUNCTIONS
+    //------------------------
+    function getPayoutInfo(uint256 _payoutId) public view payoutExists(_payoutId) returns (Payout memory) {
+        return payoutsById[_payoutId];
+    }
+
+    function getPayoutsForAsset(address _assetAddress) public view returns (Payout[] memory) {
+        uint256[] memory payoutIds = payoutsByAssetAddress[_assetAddress];
+        Payout[] memory assetPayouts = new Payout[](payoutIds.length);
+
+        for (uint i = 0; i < payoutIds.length; i++) {
+            assetPayouts[i] = payoutsById[payoutIds[i]];
+        }
+
+        return assetPayouts;
+    }
+
+    function hasClaimedFunds(uint256 _payoutId, address _wallet) public view payoutExists(_payoutId) returns (bool) {
+        return payoutClaims[_payoutId][_wallet];
+    }
+
+    //------------------------
+    //  STATE CHANGE FUNCTIONS
+    //------------------------
+    function createPayout(
+        IERC20 _asset,
+        uint256 _totalAssetAmount,
+        address[] memory _ignoredAssetAddresses,
+
+        bytes32 _assetSnapshotMerkleRoot,
+        uint256 _assetSnapshotMerkleDepth,
+        uint256 _assetSnapshotBlockNumber,
+
+        IERC20 _rewardAsset,
+        uint256 _totalRewardAmount
+    ) public returns (uint256 payoutId) {
+        // input validations
+        require(_totalAssetAmount > 0, "PayoutManager: cannot create payout without holders");
+        require(_totalRewardAmount > 0, "PayoutManager: cannot create payout without reward");
+        require(_assetSnapshotMerkleDepth > 0, "PayoutManager: Merkle tree depth cannot be zero");
+
+        address payoutOwner = msg.sender;
+
+        // verify that payout owner approved enough reward asset
+        require(_totalRewardAmount <= _rewardAsset.allowance(payoutOwner, address(this)), "PayoutManager: insufficient reward asset allowance");
+
+        // transfer reward asset
+        _rewardAsset.transferFrom(payoutOwner, address(this), _totalRewardAmount);
```

Comment:
Use transfer functions at the end. First, all require then state-changing functions and after all, comes transfer. Put this above emit event.